File: hashmap.c
#include <stdio.h>  // Standard Input/Output for functions like printf and fprintf
#include <stdlib.h> // General utilities like malloc and exit
#include "hashmap_structs.h" // Assume this file contains the definitions of HashMap and ListNode

// Function to initialize an empty hash map
void initHashMap(HashMap *map, int size) {
    map->size = size;      // Set the initial size of the hash table
    map->buckets = malloc(sizeof(ListNode*) * size); // Allocate memory for the buckets array
    
    // Error handling: check if memory allocation was successful
    if (map->buckets == NULL) {
        fprintf(stderr, "Memory allocation failed\n");  // Print error message to stderr
        exit(1);    // Exit the program with a non-zero status to indicate failure
    }

    // Initialize each bucket to NULL (empty)
    for (int i = 0; i < size; i++) {
        map->buckets[i] = NULL;
    }
}


// Hash function to calculate the index for a given key
int hash(int key, int size) {
    return key % size;  // Simple modulo hash for demonstration 
}

// Function to insert or update a key-value pair in the hash map
void put(HashMap *map, int key, int value) {
    int index = hash(key, map->size);
    ListNode *node = map->buckets[index];

    // Check if the key already exists in the hash table
    while (node != NULL) {
        if (node->key == key) {
            node->value = value; // Update the value if the key exists
            return;
        }
        node = node->next;
    }

    // Key not found, create a new node to store the key-value pair
    ListNode *newNode = malloc(sizeof(ListNode));
    
    // Error handling for memory allocation
    if (newNode == NULL) {
        fprintf(stderr, "Memory Allocation Failed");
        exit(1);
    }

    newNode->key = key;
    newNode->value = value;
    newNode->next = map->buckets[index]; // Insert at the beginning of the linked list
    map->buckets[index] = newNode;
}

// Function to retrieve the value associated with a key
int *get(HashMap *map, int key) {  // Returns a pointer to the value (allows modification)
    int index = hash(key, map->size);
    ListNode *node = map->buckets[index];

    // Search the linked list at the calculated index
    while (node != NULL) {
        if (node->key == key) {
            return &node->value; // Return the address of the value for direct access
        }
        node = node->next;
    }

    return NULL;  // Key not found
}


// Function to free the memory allocated for the hash map
void freeHashMap(HashMap *map) {
    for (int i = 0; i < map->size; i++) {
        ListNode *current = map->buckets[i];
        while (current != NULL) {
            ListNode *temp = current;
            current = current->next;
            free(temp); // Free each node in the linked list
        }
    }
    free(map->buckets); // Free the buckets array
    map->buckets = NULL; // Prevent dangling pointers
}

File: hashmap.py
class HashMap:
    """A simple implementation of a hash map using lists for buckets."""

    def __init__(self, size):
        """
        Initializes the hash map with a specified size.

        Args:
            size: The number of buckets in the hash map.
        """
        self.size = size  # Store the size for use in the hash function and bucket access
        self.buckets = [[] for _ in range(self.size)]  # Create empty lists for each bucket

    def __getitem__(self, key):
        """
        Retrieves the value associated with the given key.

        Args:
            key: The key to look up.

        Returns:
            The value associated with the key, or None if not found.
        """
        hashpoint = self.hash(key)  # Calculate the bucket index using the hash function
        bucket = self.buckets[hashpoint] # Access the correct bucket
        for k, v in bucket:  # Iterate through key-value pairs in the bucket
            if k == key:
                return v  # Return the value if the key matches
        return None  # Key not found in the bucket

    def insert(self, key, value):
        """
        Inserts a key-value pair into the hash map. If the key already exists, updates the value.

        Args:
            key: The key to insert or update.
            value: The value to associate with the key.
        """
        hashpoint = self.hash(key)
        bucket = self.buckets[hashpoint]

        for i, (k, _) in enumerate(bucket):  # Check for existing key in bucket
            if k == key:
                bucket[i] = (key, value)  # Update the value if key exists
                return  # Exit the function to avoid appending again

        bucket.append((key, value))  # Append a new key-value pair to the bucket

    def hash(self, key):
        """
        Calculates the hash value for a given key.

        Args:
            key: The key to hash.

        Returns:
            The hash value, which is an integer within the range of bucket indices.
        """
        return hash(key) % self.size  # Use Python's built-in hash function and modulo operator



